compiler/luci/pass/src/PropagateConcatenationQparamPass.cpp

```diff
@@ -0,0 +1,161 @@
+
+#include <math.h>
+
+namespace
+{
+
+bool is_uint8_quantized(luci::CircleNode *node)
+{
+  if (node->dtype() == loco::DataType::U8 && node->quantparam() != nullptr &&
+      node->quantparam()->scale.size() > 0 && node->quantparam()->zerop.size() > 0)
+    return true;
+
+  return false;
+}
+
+void overwrite_quantparam(luci::CircleConcatenation *concat, luci::CircleNode *target)
+{
+  auto concat_qparam = concat->quantparam();
+  assert(concat_qparam != nullptr);
+
+  auto target_qparam = target->quantparam();
+  assert(target_qparam != nullptr);
+  target_qparam->min = concat_qparam->min;
+  target_qparam->max = concat_qparam->max;
+  target_qparam->scale = concat_qparam->scale;
+  target_qparam->zerop = concat_qparam->zerop;
+  target_qparam->quantized_dimension = concat_qparam->quantized_dimension;
+}
+
+} // namespace
+
+namespace luci
+{
+
+void propagate_concat_quantparam(luci::CircleConcatenation *concat)
+{
+  // Check if concat is uint8-quantized
+  if (!is_uint8_quantized(concat))
+    return;
+
+  // Check if concat has no fused activation function
+  if (concat->fusedActivationFunction() != luci::FusedActFunc::NONE)
+    return;
+
+  const auto num_inputs = concat->numValues();
+  for (uint32_t i = 0; i < num_inputs; i++)
+  {
+    auto node = loco::must_cast<luci::CircleNode *>(concat->arg(i));
+
+    // Skip if this input is CONCAT Op
+    if (node->opcode() == luci::CircleOpcode::CONCATENATION)
+      continue;
+
+    // Skip if this input is used by other Ops
+    auto succs = loco::succs(node);
+    if (succs.size() != 1)
+      continue;
+
+    assert(succs.find(concat) != succs.end());
+
+    // Quantize constant values
+    if (node->opcode() == luci::CircleOpcode::CIRCLECONST)
+    {
+      luci::CircleConst *const_node = loco::must_cast<luci::CircleConst *>(node);
+      assert(const_node->dtype() == loco::DataType::FLOAT32);
+      if (const_node->dtype() != loco::DataType::FLOAT32)
+        throw std::runtime_error("Unsupported data type for constant input of concatenation Op");
+
+      uint32_t size = const_node->size<loco::DataType::FLOAT32>();
+
+      const auto concat_qparam = concat->quantparam();
+      assert(concat_qparam->scale.size() == 1);
+
+      const auto scaling_factor = concat_qparam->scale[0];
+      const auto zerop = concat_qparam->zerop[0];
+
+      const float scaling_factor_inv = 1.0 / scaling_factor;
+      std::vector<int32_t> quantized_values(size);
+      for (uint32_t i = 0; i < size; ++i)
+      {
+        auto data = const_node->at<loco::DataType::FLOAT32>(i);
+        quantized_values[i] = static_cast<int32_t>(std::round(data * scaling_factor_inv) + zerop);
+      }
+
+      const_node->dtype(loco::DataType::U8);      // change the type of tensor
+      const_node->size<loco::DataType::U8>(size); // resize tensor
+      for (uint32_t i = 0; i < size; ++i)
+      {
+        const_node->at<loco::DataType::U8>(i) = std::min(255, std::max(0, quantized_values[i]));
+      }
```

Comment:
IMHO It's worth extracting into a function.

Reply:
It would enhance readability. I've updated the code.